### PR TITLE
Refactored CloudFormation Template template

### PR DESCRIPTION
**Pull request by automatthew**
Refactored `mixins/api.yaml` and the config-preprocessors that support it so that the only logic done in Handlebars template language is conditionals and iteration.  This makes it much easier to understand and alter the api.yaml template.

This should produce the exact same CloudFormation templates as before, with one exception:  It uses the "title" property of the supplied schemas to name the API Gateway Model, rather than reflecting on the key by which you reach the schema in the Panda Sky API Description.

I tested this by saving a copy of the CloudFormation template for Blurb9, then asserting that the newly generated CloFoTem was deepEqual to the known good one.

There are some additional elements I want to rework later, but this PR is for a strict refactor (barring that one exception).

### CODE REVIEW

**Comment by automatthew** [comment]
Here's the known good template I was targeting:

https://github.com/pandastrike/panda-sky/blob/api-template-refactor/test/data/blurb9/_cloudformation.yaml

And the junky unit test:

https://github.com/pandastrike/panda-sky/blob/api-template-refactor/test/unit/cloudformation.coffee

**Comment by automatthew** [comment]
One of the interesting and useful side effects of this work is that we can now dump as YAML the `config` object that we use to populate the mixins/api.yaml template.  This allows us more easily to troubleshoot problems in the structure of the template, as well as the formatting of the configuration data with which we propose to fill it.

The current config data looks like this as yaml:

https://github.com/pandastrike/panda-sky/blob/api-template-refactor/test/data/blurb9/_api_config.yaml

**Comment by automatthew** [comment]
This also gets us closer to the ideas behind Emerald Robot.

In rising order of ambition:

1. a convenient library and CLI for merging YAML config files and stitching them into Handlebars template files.
2. a clean, opinionated library and CLI for working with the AWS CloudFormation API, focusing on use of CloFoTemps
3. easy-to-use "mixins": Handlebars templates for various CloudFormation Resource setups.
4. schematization of the configs required to fulfill the mixins we supply.

**Comment by automatthew** [comment]
One of the things I'm considering doing with Panda Sky to make it easier to troubleshoot is to create a `_sky_build` directory and write YAML dumps there for things like the API-template config data, and the CloudFormation template, and whatever other configuration structures Sky uses.

File: src/configuration/preprocessors/api/actions.coffee [hunk 1]
```diff
@@ -21,13 +22,38 @@ module.exports = (description) ->
...
+
   for r, resource of resources
     methods = ["options"]
     for a, action of resource.actions
+      action.method = capitalize action.method
       methods.push action.method
-      resources[r]["actions"][a].camelName = makeCamelName r, action
-      resources[r]["actions"][a].dashName = makeDashName r, action
-      resources[r]["actions"][a]["signature"].dependent = hasDependency action.signature
+      action.camelName = makeCamelName r, action
```
Comment (dyoder) [inline src/configuration/preprocessors/api/actions.coffee]: This all seems like a giant step in the right direction. I wanted to see how far I could get using just the API description and the template, but the result is almost unmaintainable. The trick is figuring out what these things are (ex: gateway method name) but I think this does a pretty job of doing that.

File: src/configuration/preprocessors/api/index.coffee [hunk 2]
```diff
@@ -1,32 +1,36 @@
...
-  description = yield extractResources description
+  mungedConfig = yield extractResources mungedConfig
 
   # Compute the formatted template names for API action defintions.
-  description = yield extractActions description
+  mungedConfig = yield extractActions mungedConfig
 
   # Add the possible HTTP responses to every API action specification.
-  description = yield addResponses description
+  mungedConfig = yield addResponses mungedConfig
```
Comment (dyoder) [inline src/configuration/preprocessors/api/index.coffee]: this repeated modification is clunky. might be a good place for `go`:

`go (extractParameters description),  addVariables, extractCFr, extractResources, extractActions, addResponses`

#nit

Comment (dyoder) [inline src/configuration/preprocessors/api/index.coffee]: also feels a bit like some larger abstraction is missing, like those steps should be scoped within something, ex:

```coffee
{parameters, template, resources, actions} = extractors
{variables, response} = decorators
go [
  (paramaters description),
  variables,
  template,
  resources,
  actions,
  responses]
```

#anotherNit

File: src/configuration/preprocessors/api/index.coffee [hunk 3]
```diff
@@ -1,32 +1,36 @@
...
 
   # Select the runtime for the Lambda, setting a default if not set.
-  description = yield selectRuntime description
-
-  # Add environment varialbles that are injected into every Lambda.
-  description = yield addVariables description
+  mungedConfig = yield selectRuntime mungedConfig
```
Comment (dyoder) [inline src/configuration/preprocessors/api/index.coffee]: why is that async?

**Comment by dyoder** [comment]
Can we just use a build directory specified by the the `sky.yaml` config instead of hardcoding it?

**Pull request merged**